### PR TITLE
Add possiblity to send files in request body

### PR DESCRIPTION
**Pull request by daanvdk**
Adds a possibility to send files deeply nested within data.

To do this you have to send a `multipart/form-data` request with the data encoded as json in a field called `data`. The files within this data have to be replaced with `null` and added to the multipart data as a file, the name of this file should be `file:<path>` where `<path>` is the keys that lead to the file joined with `.`.

Some other notable changes: So because of this I wanted to load json that is a str instead of only bytes, so I removed the `.decode()` in jsonloads. This is backwards compatible for python >=3.6 since in that release `json.loads` was changed to also accept binary data.

### CODE REVIEW

File: binder/json.py [hunk 1]
```diff
@@ -48,14 +48,12 @@ def jsondumps(o, default=default, indent=None):
 	return json.dumps(o, default=default, indent=indent)
 
 
-
 def jsonloads(data):
 	try:
-		return json.loads(data.decode())
+		return json.loads(data)
```
Comment (stefanmajoor) [inline binder/json.py:53]: jsonloads is also used a lot on standalone. Will this potentially break all that usage?

Comment (daanvdk) [reply binder/json.py:53]: As addressed in the PR description: depends on the python version.
Since python 3.6 `json.loads` accepts both `str` and `bytes` so the `decode` is bultin when its `bytes`. With earlier versions that would indeed break, but according to the setup.py those are not supported anyway.

File: binder/views.py [hunk 2]
```diff
@@ -65,6 +67,32 @@ def get_default_annotations(model):
 	return annotations
 
 
+def split_path(path):
```
Comment (stefanmajoor) [inline binder/views.py:70]: Can you add some sort description of what split and join path actually do? And add the type hinting  the arguments? That will make ti much easier to understand what is going on

File: binder/views.py [hunk 3]
```diff
@@ -481,12 +509,11 @@ def _get_objs(self, queryset, request, annotations=None):
 
 			data = {}
 			for f in fields:
-				if isinstance(f, models.fields.files.FileField):
+				if isinstance(f, models.FileField):
 					file = getattr(obj, f.attname)
-					if file:
+					if file.name:
```
Comment (stefanmajoor) [inline binder/views.py:514]: What is the reasoning behind this change? Does it have any side effects?

File: binder/views.py [hunk 4]
```diff
@@ -1422,7 +1449,10 @@ def store_m2m_field(obj, field, value, request):
 
 		# Skip re-fetch and serialization via get_objs if we're in
 		# multi-put (data is discarded!).
-		if getattr(request, '_is_multi_put', False):
+		if (
+			getattr(request, '_is_multi_put', False) or
+			getattr(request, '_is_file_upload', False)
```
Comment (stefanmajoor) [inline binder/views.py:1454]: Does this mean we do not get the response in json anymore of add files to our request. Is there any advantage in not doing it?

File: binder/views.py [hunk 5]
```diff
@@ -1614,6 +1645,97 @@ def _store_field(self, obj, field, value, request, pk=None):
...
+				elif isinstance(f, models.FileField):
+					if not isinstance(value, File) and value is not None:
+						raise BinderFieldTypeError(self.model.__name__, field)
+
+					if value is not None:
+						if value.size > self.max_upload_size * 10**6:
+							raise BinderFileSizeExceeded(self.max_upload_size)
+
+						if getattr(f, 'allowed_extensions', None) is not None:
+							extension = None if '.' not in value.name else value.name.split('.')[-1]
```
Comment (stefanmajoor) [inline binder/views.py:1658]: We should use [os.path.splitext](https://docs.python.org/3/library/os.path.html#os.path.splitext), instead of reinventing the wheel all the time.

**Review by stefanmajoor — CHANGES_REQUESTED** [review]
The idea is good, and it is getting there. I do miss two things:

- Documentation. Both in the code (describe what new methods do, inline comments) as well as general documentation (this is how the method works). I had to try to figure out what exactly you were trying to implement from the implementation
- Only the good weather scenario is tested for the file upload. Can  you also add tests  for the edge cases? e.g. what happens if you do not set the "Data", or the "data" is incorrect. O

File: binder/views.py (hunk 4 above)
Comment (daanvdk) [reply binder/views.py]: This check is about file uploads through dispatch file field, not the new file uploads. Added a comment to clarify.

File: binder/views.py (hunk 5 above)
Comment (daanvdk) [reply binder/views.py:1678]: If this was newly written code I agree. However in this case it was just moved from dispatch_file_field to store_field, in which case I think its more important to remain 100% backwards compatible.

File: binder/views.py [hunk 6]
```diff
@@ -67,6 +68,42 @@ def get_default_annotations(model):
 	return annotations
 
 
+def split_path(path):
+	"""
+	This function splits a dot seperated path into a list of keys.
```
Comment (stefanmajoor) [inline binder/views.py:73]: I still don't understand what this is doing, and why we need this function in the first place. This still seems like something which can be done in a regex.

Can you give some examples in the documentation?

Comment (stefanmajoor) [reply binder/views.py:73]: Even even better, can you add some unittests for this methods? Then we know for sure it is working as well

File: binder/views.py (hunk 5 above)
Comment (stefanmajoor) [reply binder/views.py:1698]: I disagree, since the only reason I can think that it is not 100% backwards compatible, is if the code gives a wrong extension. That would be a security issue, and a bug that needs to be fixed

File: binder/views.py [hunk 7]
```diff
@@ -1626,6 +1666,116 @@ def _store_field(self, obj, field, value, request, pk=None):
...
+
+						if getattr(f, 'allowed_extensions', None) is not None:
+							extension = None if '.' not in value.name else value.name.split('.')[-1]
+
+							if extension not in f.allowed_extensions:
+								raise BinderFileTypeIncorrect([{'extension': t} for t in f.allowed_extensions])
+
+						if isinstance(f, (models.ImageField, BinderImageField)):
+							try:
+								img = Image.open(value)
```
Comment (stefanmajoor) [inline binder/views.py:1705]: Could you add  that if we have an ImageField, and no allowed extensions are specified,  that we check on the extension before parsing the image? This is a security issue otherwise

File: binder/views.py [hunk 8]
```diff
@@ -1626,6 +1666,116 @@ def _store_field(self, obj, field, value, request, pk=None):
...
+							except Exception:
+								raise BinderImageError('Could not parse the file as an image.')
+
+							format = img.format.lower()
+							if not format in ('png', 'gif', 'jpeg'):
+								raise BinderFileTypeIncorrect([{'extension': t, 'mimetype': 'image/' + t} for t in ['jpeg', 'png', 'gif']])
+
+							width, height = img.size
+							if format == 'jpeg':
```
Comment (stefanmajoor) [inline binder/views.py:1714]: For readability's sake, would it be possible to break down the very, very large method into some well named smaller methods?  We are going 8 levels deep inside in a method which goes on and on and on.

**Review by stefanmajoor — CHANGES_REQUESTED** [review]
